Blog post on the outbox pattern

gunnarmorling opened the pull request:
Accompanying source code: https://github.com/debezium/debezium-examples/tree/master/outbox

~~@jpechane, I'm still going to refine this one, but here's the first draft. Any feedback on the overall flow would be highly welcomed.~~

Update: pretty happy with it now :)

jpechane commented on blog/2019-02-19-reliable-microservices-data-exchange-with-the-outbox-pattern.adoc:
```diff
@@ -0,0 +1,598 @@
+One general issue of such a synchronous approach is that it lacks re-playability,
+i.e. the possibility for new consumers to arrive after events have been sent and still be able to consume the entire event stream from the beginning.
+
+This can be achieved by using an asynchronous data exchange approach:
```
I think this is not fully correct - this implies that `asynchronous data exchange` == `replayabilty` but those two concepts overlap only partially.

jpechane commented on blog/2019-02-19-reliable-microservices-data-exchange-with-the-outbox-pattern.adoc:
```diff
@@ -0,0 +1,598 @@
+and without touching the source service.
+E.g. consider a data warehouse which should keep information about all the orders ever placed, or some full-text search functionality on orders based on https://www.elastic.co/products/elasticsearch[Elasticsearch].
+Once the purchase order events are in a Kafka topic
+(and depending on the topic's retention policy, they could remain their for an arbitrary long time or even indefinitely),
```
Should be `there` not `their`

gunnarmorling replied:
Thx.

jpechane commented on blog/2019-02-19-reliable-microservices-data-exchange-with-the-outbox-pattern.adoc:
```diff
@@ -0,0 +1,598 @@
+
+== The Issue of Dual Writes
+
+In order to provide their functionality, data-centered microservices will typically have their own local database.
```
`local database` or `local data store`?

gunnarmorling replied:
Both is fine I guess. But I like store better, too, changing it :)

jpechane commented on blog/2019-02-19-reliable-microservices-data-exchange-with-the-outbox-pattern.adoc:
```diff
@@ -0,0 +1,598 @@
+this may cause no shipment to be created for a seemingly successfully placed order.
+Or a shipment gets created, but then there'd be no trace about the corresponding purchase order in the order service itself.
+
+So how can this situation be avoided?
```
I think a short mention of XA and why it was not chosen should be added.

gunnarmorling replied:
Ok, I'll try. But "not chosen" sounds as if it were an option; but AFAICS it isn't, as Kafka cannot be enlisted in an XA transcation.

gunnarmorling replied:
Done.

jpechane commented on blog/2019-02-19-reliable-microservices-data-exchange-with-the-outbox-pattern.adoc:
```diff
@@ -0,0 +1,598 @@
+Value: {"eventType":"OrderLineUpdated","ts_ms":1550308226963,"payload":"{\"orderId\": 4, \"newStatus\": \"CANCELLED\", \"oldStatus\": \"ENTERED\", \"orderLineId\": 7}"}
+----
+
+The `payload` field with the message values is the string-ified JSON representation of the original events.
```
The article does not explain how the sringification was achieved.

gunnarmorling replied:
Clarified it.

jpechane commented on blog/2019-02-19-reliable-microservices-data-exchange-with-the-outbox-pattern.adoc:
```diff
@@ -0,0 +1,598 @@
+
+Its columns are these:
+
+* `id`: unique id of each message; can be used by consumers to detect any duplicate events, e.g. when restarting to read messages after a failure.
```
How is the `id` generated?

gunnarmorling replied:
Upon event instantiation. Clarified it.

sherl0cks commented on blog/2019-02-19-reliable-microservices-data-exchange-with-the-outbox-pattern.adoc:
```diff
@@ -0,0 +1,701 @@
+but when looking at the contents of the table itself, it will always be empty.
+This means that no additional disk space is needed for the table
+(apart from the log file elements which will automatically be discarded at some point)
+and also no separate house-keeping process is required to stop it from growing indefinitely.
```
Hmmm - I have very mixed feelings about this. This is related to the idea that events live in Kafka long term. Essentially it moves operational complexity from a well known industry space (relational DBs) towards a very poorly known industry space (kafka / k8s / ocp).

There are other complexities here too. I need to think about this.

sherl0cks replied:
Per the previous comment thread - I'm OK with this approach, but we should add a line here that mentions that this approach is a corollary of the indefinite storage in Kafka (which again is one option of several). I'm just worried that I'll get customers telling me "This is the one and only way to implement outbox! Gunnar said so!"

gunnarmorling commented on blog/2019-02-19-reliable-microservices-data-exchange-with-the-outbox-pattern.adoc:
```diff
@@ -0,0 +1,701 @@
+
+== The Issue of Dual Writes
+
+In order to provide their functionality, data-centered microservices will typically have their own local data store.
```
Interesting, @jpechane was stumbling upon it, too. I'll remove it.  I had it there to make a distinction to microservices that don't need a data store at all. But it's not too important.

gunnarmorling commented on blog/2019-02-19-reliable-microservices-data-exchange-with-the-outbox-pattern.adoc:
```diff
@@ -0,0 +1,701 @@
+Or, the other way around, we might have sent the message to Kafka but failed to persist the purchase order in the local database.
+Both situations are undesirable;
+this may cause no shipment to be created for a seemingly successfully placed order.
+Or a shipment gets created, but then there'd be no trace about the corresponding purchase order in the order service itself.
```
In which case could a re-ordering occur?

sherl0cks replied:
The more I think about it, the more this is a corner case not worth exploring. Your previous points clearly communicate the challenge. Closing this comment.

gunnarmorling commented on blog/2019-02-19-reliable-microservices-data-exchange-with-the-outbox-pattern.adoc:
```diff
@@ -0,0 +1,701 @@
+
+* `id`: unique id of each message; can be used by consumers to detect any duplicate events, e.g. when restarting to read messages after a failure.
+Generated when creating a new event.
+* `aggregatetype`: the type of the aggregate root to which a given event is related; this could for instance be "purchase order" or "customer".
```
Good point, I'll add some explanatory sentence before that. That said, I don't see the term "aggregate" strictly tied to DDD, and mind you, last time I actually made the link, I got called out :) But in any case it's worth being explained upfront.

hpgrahsl replied:
@gunnarmorling given what happended last time in our joint article I'd suggest that if you really want to refer to the DDD idea of an aggregate you might want to reference Eric Evans?

sherl0cks replied:
@gunnarmorling if you don't want to address this as a DDD idea, thats 100% OK. But you should provide context to what you mean specifically. If you do want to address it as a DDD idea, then I think I know how to word it properly and I can send a PR if you would like it.

gunnarmorling replied:
I've added some explanation now; @sherl0cks, do you think it's ok as such now? If not, a PR would be much appreciated.

gunnarmorling commented on blog/2019-02-19-reliable-microservices-data-exchange-with-the-outbox-pattern.adoc:
```diff
@@ -0,0 +1,701 @@
+E.g. consider a data warehouse which should keep information about all the orders ever placed, or some full-text search functionality on purchase orders based on https://www.elastic.co/products/elasticsearch[Elasticsearch].
+Once the purchase order events are in a Kafka topic
+(and depending on the topic's retention policy, they could remain there for an arbitrary long time, or even indefinitely),
+new consumers can subscribe, process the topic from the very beginning and materialize a view of all the data in a microservice's database, search index, data warehouse etc.
```
I see; do you have an alternative for replayability? Also I'm curious, that thread you linked is nearly three years old, perhaps things have changed since then? I've seen references to long retention, but I'm sure what's the current agreed on thinking on that one is.

hpgrahsl replied:
while it is right that very long (indefinite) storage brings its own challenges there are real-world examples where this is done. the "most famous" one that quickly comes to my mind is the NYT approach towards permanent storage https://www.confluent.io/blog/publishing-apache-kafka-new-york-times/ and a referring post https://www.confluent.io/blog/okay-store-data-apache-kafka/

sherl0cks replied:
From the NYT blog:
```
Because the topic is single-partition, it needs to be stored on a single disk, due to the way Kafka stores partitions. This is not a problem for us in practice, since all our content is text produced by humans — our total corpus right now is less than 100GB, and disks are growing bigger faster than our journalists can write.
```
So that's 100GB for > 150 years of data. I think many, many use cases customers in production will have much greater different scale than  .66 GB / year and design systems to use more than one partition. 

I'm not saying that the approach is invalid, I'm just saying that proposing this sort of architecture (indefinite storage in the Kafka cluster, no event storage in the DB) has serious operational implication and at minimal we should highlight this fact so readers are aware of trade off & alternative.

As to reply ability, I think the most likely case is going to after an outage where we've used traditional message queueing for ages. The idea of a whole new service replaying 5 years of message history from multiple services to create an initial state is not a daily or weekly requirement at most organizations. If the requirement does come up, I would prefer to leverage an event log in my RDBMS and create a one off debezium connector for this specific purpose. This gets in to the whole idea of optionality I write about [here](https://medium.com/@justinmichaelholmes/creating-optionality-with-domain-event-logging-7eb8f1f422e7)

Again - these alternatives. Neither is right or wrong and all I am asking is that we highlight this approach as one of many.

sherl0cks commented on blog/2019-02-19-reliable-microservices-data-exchange-with-the-outbox-pattern.adoc:
```diff
@@ -0,0 +1,701 @@
+In order to "send" events to the outbox, code in the order service could in general just do an `INSERT` into the outbox table.
+However, it's a good idea to go for a slightly more abstract API, allowing to adjust implementation details of the outbox later on more easily, if needed.
+https://docs.jboss.org/weld/reference/latest/en-US/html/events.html[CDI events] come in very handy for this.
+They can be raised in the application code and will be processed _synchronously_ by the outbox event sender,
```
May be worth mentioning that a Spring impl is quite similar. https://zoltanaltfatter.com/2017/06/09/publishing-domain-events-from-aggregate-roots/ and https://github.com/haithamshahin333/Domain-Event-Sample

gunnarmorling replied:
Added a reference to `@DomainEvents`.

sherl0cks commented on blog/2019-02-19-reliable-microservices-data-exchange-with-the-outbox-pattern.adoc:
```diff
@@ -0,0 +1,701 @@
+At the same time, the pattern enables asynchronous event propagation to other microservices.
+Apache Kafka acts as a highly scalable and reliable backbone for the messaging amongst the services.
+Given the right topic retention settings, new consumers may come up long after an event has been originally produced,
+and build up their own local state based on the event history.
```
This is again subject to the same "there are options here" qualification.

gunnarmorling replied:
Should be addressed now.

sherl0cks commented on blog/2019-02-19-reliable-microservices-data-exchange-with-the-outbox-pattern.adoc:
```diff
@@ -0,0 +1,701 @@
+One last thing to keep in mind is that the structure of the events exposed via the outbox should be considered a part of the emitting service's API.
+I.e. when needed, their structure should be adjusted carefully and with compatibility considerations in mind.
+This is to ensure to not accidentally break any consumers when upgrading the producing service.
+At the same time, consumers should be lenient when handling messages and for instance not fail when encountering unknown attributes within received events.
```
This last point is really important, and its the key corollary to "messaging makes you more decoupled." It may be worth making this point earlier in the blog.

gunnarmorling replied:
I'm touching on it earlier on already (and expanded that one a bit):

> By explicitly emitting events via records in the outbox table,
the application code can work with events are structured in a way suitable for external consumers.
> This also ensures that event consumers won't break
when for instance altering the internal domain model or the `PurchaseOrder` table.

I think it's ok that way.

gunnarmorling commented on blog/2019-02-19-reliable-microservices-data-exchange-with-the-outbox-pattern.adoc:
```diff
@@ -0,0 +1,697 @@
+But it makes sense when remembering how log-based CDC works:
+it doesn't examine the actual contents of the table in the database, but instead it tails the append-only transaction log.
+The calls to `persist()` and `remove()` will create an `INSERT` and a `DELETE` entry in the log once the transaction commits.
+After that, Debezium will process these events (handling the `INSERT` and ignoring the `DELETE`).
```
Good idea; done.

gunnarmorling commented:
Hey @sherl0cks, so first of all, thank you (and everyone else!) to all your comments, they are very much appreciated.

In particular the discussion around indefinite topic retention is a good one. This is presented in a slightly different angle now, please see the two note boxes "Dealing with Topic Growth" and "Deletion of Events from Kafka Topics" for the details. This now discusses compacting the event topics (so you'd only keep the last event pertaining to a given purchase order, customer etc., which requires that each event fully describes the represented aggregate, and not just the "delta") and also the idea of moving historic events to cold storage such as an S3 bucket. Could you perhaps review the last commit I've added and let me know whether you see any remaining issues? From my perspective, it's good to go and I hoped I've addressed all your remarks. Thanks again so much!

gunnarmorling commented:
And applied; should be live in a few minutes. Thanks again everyone for all your comments, that helped a lot!
